src/BD_sensor.c

```diff
@@ -0,0 +1,541 @@
+
+///////////BDsensor as endstop
+struct timer time_bd;
+#include "autoconf.h"
+struct endstop {
+    struct timer time;
+    uint32_t rest_time, sample_time, nextwake,pin_num;
+    struct trsync *ts;
+    uint8_t flags, sample_count, trigger_count, trigger_reason;
+    struct gpio_in pin;
+};
+
+enum { ESF_PIN_HIGH=1<<0, ESF_HOMING=1<<1 };
+static uint_fast8_t endstop_oversample_event(struct timer *t);
+static struct endstop e ;
+///////////////
+
+int z_index=0;
+
+enum { POSITION_BIAS=0x40000000 };
+
+struct stepper_move {
+    struct move_node node;
+    uint32_t interval;
+    int16_t add;
+    uint16_t count;
+    uint8_t flags;
+};
+
+struct stepper {
+    struct timer time;
+    uint32_t interval;
+    int16_t add;
+    uint32_t count;
+    uint32_t next_step_time, step_pulse_ticks;
+    struct gpio_out step_pin, dir_pin;
+    uint32_t position;
+    struct move_queue_head mq;
+    struct trsync_signal stop_signal;
+    // gcc (pre v6) does better optimization when uint8_t are bitfields
+    uint8_t flags : 8;
+};
+
+struct step_adjust{
+    uint32_t cur_z;
+    int adj_z_range;
+    int invert_dir;
+    int steps_per_mm;
+    int step_time;
+    int zoid;//oid for all the z stepper
+};
+
+#define NUM_Z_MOTOR  6
+struct step_adjust step_adj[NUM_Z_MOTOR];//x,y,z
+
+struct _step_probe{
+    int min_x;
+    int max_x;
+    int points;
+    int steps_at_zero;
+    int steps_per_mm;
+    int xoid;//oid for x stepper
+    int x_count;
+    int x_dir;
+    int y_dir;
+    int kinematics;//0:cartesian,1:corexy,2:delta
+    ////for y
+    int min_y;
+    int max_y;
+    int y_steps_at_zero;
+    int y_steps_per_mm;
+    int y_oid;//oid for y stepper
+    ////
+    int x_data[64];
+};
+
+struct _step_probe stepx_probe;
+
+
+void BD_i2c_write(unsigned int addr);
+uint16_t BD_i2c_read(void);
+
+int BD_i2c_init(uint32_t _sda,uint32_t _scl,
+    uint32_t delays,uint32_t h_pose,int z_adjust)
+{
+    int i=0;
+    sda_pin=_sda;
+    scl_pin =_scl;
+    homing_pose = h_pose;
+    z_ofset = z_adjust;
+    if (z_ofset > 500)
+        z_ofset = 0;
+    if(delays>0)
+        delay_m=delays;
+    sda_gpio=gpio_out_setup(sda_pin, 1);
+    scl_gpio=gpio_out_setup(scl_pin, 1);
+
+    gpio_out_write(sda_gpio, 1);
+    gpio_out_write(scl_gpio, 1);
+    for (i=0;i<NUM_Z_MOTOR;i++){
+        step_adj[i].cur_z=0;
+        step_adj[i].zoid=0;
+        step_adj[i].adj_z_range=0;
+    }
+    stepx_probe.xoid=0;
+    stepx_probe.y_oid=0;
+    BD_i2c_write(1022); //reset BDsensor
+    return 1;
+}
+
+uint32_t nsecs_to_ticks_bd(uint32_t ns)
+{
+    return ns * (CONFIG_CLOCK_FREQ / 1000000);
+}
+
+
+void ndelay_bd_c(uint32_t nsecs)
+{
+    if (CONFIG_MACH_AVR)
+        return;
+    uint32_t end = timer_read_time() + nsecs_to_ticks_bd(nsecs);
+    while (timer_is_before(timer_read_time(), end))
+        irq_poll();
+}
+void ndelay_bd(uint32_t nsecs)
+{
+    int i=1;
+    while(i--)
+        ndelay_bd_c(nsecs);
+}
+
+unsigned short BD_Add_OddEven(unsigned short byte)
```

Comment:
a little bit confusing to name a "short" `byte` =)) was wondering how a byte can have more than 8 (10) bits looking at the loop :)